pandas/core/frame.py

```diff
@@ -7926,8 +7941,12 @@ def _reduce(
 
         assert filter_type is None or filter_type == "bool", filter_type
 
-        dtype_is_dt = self.dtypes.apply(
-            lambda x: is_datetime64_any_dtype(x) or is_period_dtype(x)
+        dtype_is_dt = np.array(
+            [
+                is_datetime64_any_dtype(values.dtype) or is_period_dtype(values.dtype)
+                for values in self._iter_arrays()
+            ],
+            dtype=bool,
```

Comment:
the perf issue here is in the `self.dtypes` call?

Reply:
as much as im trying to avoid `._data`, might be cleaner to use `self._data.get_dtypes` here

Reply:
> the perf issue here is in the self.dtypes call?

Both the `self.dtypes` that creates a Series, as the `apply` that creates another Series with inference. 

> as much as im trying to avoid ._data, might be cleaner to use self._data.get_dtypes here

Indeed, that would be even more appropriate here. 

Now, I am happy to change it to that, but that's not really the core of this PR. I mainly wanted to add one useful case as an illustration, but mainly want the helper function for other PRs. I am happy to merge it without already using it somewhere as well.

Reply:
since we're just going to call `.any()` on this anyway, could do the any up-front on the iterator, no big deal